### PR TITLE
New Adapter: ReVantage

### PR DESCRIPTION
**Pull request by v0idxyz**
### 🔧 Type of changes
- [x] new bid adapter
- [ ] bid adapter update
- [ ] new feature
- [ ] new analytics adapter
- [ ] new module
- [ ] module update
- [ ] bugfix
- [ ] documentation
- [ ] configuration
- [ ] dependency update
- [ ] tech debt (test coverage, refactorings, etc.)

### ✨ What's the context?
What's the context for the changes?

### 🧠 Rationale behind the change
Why did you choose to make these changes? Were there any trade-offs you had to consider?

### 🔎 New Bid Adapter Checklist
- [x] verify email contact works
- [x] NO fully dynamic hostnames
- [x] geographic host parameters are NOT required
- [x] direct use of HTTP is prohibited - *implement an existing Bidder interface that will do all the job*
- [x] if the ORTB is just forwarded to the endpoint, use the generic adapter - *define the new adapter as the alias of the generic adapter*
- [x] cover an adapter configuration with an integration test

### 🧪 Test plan
How do you know the changes are safe to ship to production?

### 🏎 Quality check
- [x] Are your changes following [our code style guidelines](https://github.com/prebid/prebid-server-java/blob/master/docs/developers/code-style.md)?
- [x] Are there any breaking changes in your code?
- [x] Does your test coverage exceed 90%?
- [x] Are there any erroneous console logs, debuggers or leftover code in your changes?

### CODE REVIEW

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+ * <pre>
+ * {
+ *   "feedId": "...",
+ *   "bidder": {
+ *     "placementId": "...",   // optional
+ *     "publisherId": "..."    // optional
+ *   }
+ * }
+ * </pre>
+ */
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java]: No need for this comment

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+
+    public RevantageBidder(String endpointUrl, JacksonMapper mapper) {
+        this.endpointUrl = HttpUtil.validateUrl(Objects.requireNonNull(endpointUrl));
+        this.mapper = Objects.requireNonNull(mapper);
+    }
+
+    @Override
+    public Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest request) {
+        final List<BidderError> errors = new ArrayList<>();
+        // LinkedHashMap preserves first-seen feedId order for deterministic output.
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java]: No need for this comment

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+        final Map<String, List<Imp>> impsByFeed = new LinkedHashMap<>();
+
+        for (Imp imp : request.getImp()) {
+            try {
+                final ExtImpRevantage ext = parseImpExt(imp);
+                final String feedId = StringUtils.trimToNull(ext.getFeedId());
+                if (feedId == null) {
+                    throw new PreBidException("imp %s: missing required param feedId".formatted(imp.getId()));
+                }
+                final Imp rewrittenImp = rewriteImpExt(imp, feedId, ext);
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java]: `rewriteImpExt` -> `updateImp`, `rewrittenImp` -> `updatedImp`

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+        return Result.of(requests, errors);
+    }
+
+    private ExtImpRevantage parseImpExt(Imp imp) {
+        try {
+            final ExtPrebid<?, ExtImpRevantage> wrapper =
+                    mapper.mapper().convertValue(imp.getExt(), REVANTAGE_EXT_TYPE_REFERENCE);
+            if (wrapper == null || wrapper.getBidder() == null) {
+                throw new PreBidException("imp %s: missing imp.ext.bidder".formatted(imp.getId()));
+            }
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java]: No need for this checks. At this stage, neither the `wrapper` nor the `wrapper.getBidder()` can be `null`

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        }
+        if (StringUtils.isNotBlank(ext.getPublisherId())) {
+            bidderNode.put("publisherId", ext.getPublisherId());
+        }
+
+        final ObjectNode rewritten = mapper.mapper().createObjectNode();
+        rewritten.put("feedId", feedId);
+        rewritten.set("bidder", bidderNode);
+
+        return imp.toBuilder().ext(rewritten).build();
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java]:
```
        final ObjectNode newExt = mapper.mapper().createObjectNode();
        newExt.put("feedId", feedId);

        final ObjectNode bidderNode = rewritten.putObject("bidder");

        final String placementId = ext.getPlacementId();
        if (StringUtils.isNotBlank(placementId)) {
            bidderNode.put("placementId", placementId);
        }

        final String publisherId = ext.getPublisherId();
        if (StringUtils.isNotBlank(publisherId)) {
            bidderNode.put("publisherId", publisherId);
        }

        return imp.toBuilder().ext(newExt).build();
```

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+    }
+
+    private HttpRequest<BidRequest> buildHttpRequest(BidRequest request, String feedId) {
+        final String uri = endpointUrl + "?feed=" + URLEncoder.encode(feedId, StandardCharsets.UTF_8);
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java]: Use `HttpUtil.encodeUrl` instead

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+
+        final MultiMap headers = HttpUtil.headers();
+        headers.set(HttpUtil.ACCEPT_HEADER, "application/json");
+
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java]: No need to set `Accept` header. It's already included in `HttpUtil.headers()`

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+        return HttpRequest.<BidRequest>builder()
+                .method(HttpMethod.POST)
+                .uri(uri)
+                .headers(headers)
+                .body(mapper.encodeToBytes(request))
+                .impIds(collectImpIds(request))
+                .payload(request)
+                .build();
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java]: Use `BidderUtil.defaultRequest` instead

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+            final BidResponse response = mapper.decodeValue(httpCall.getResponse().getBody(), BidResponse.class);
+            return Result.withValues(extractBids(response, bidRequest));
+        } catch (DecodeException | PreBidException e) {
+            return Result.withError(BidderError.badServerResponse(e.getMessage()));
+        }
+    }
+
+    private static List<BidderBid> extractBids(BidResponse response, BidRequest request) {
+        if (response == null || CollectionUtils.isEmpty(response.getSeatbid())) {
+            return List.of();
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java]: `List.of` -> `Collections.emptyList`

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+                case 1:
+                    return BidType.banner;
+                case 2:
+                    return BidType.video;
+                default:
+                    // fall through
+            }
+        }
+
+        final JsonNode ext = bid.getExt();
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java]: Refactor this method:
```
    private static BidType resolveBidType(Bid bid, List<Imp> imps) {
        return bidTypeFromMtype(bid.getMtype())
                .or(() -> bidTypeFromExt(bid.getExt()))
                .or(() -> bidTypeFromAdm(bid.getAdm()))
                .or(() -> bidTypeFromImp(bid.getImpid(), imps))
                .orElseThrow(() -> new PreBidException(
                        "Cannot determine media type for bid %s on imp %s".formatted(bid.getId(), bid.getImpid())));
    }

    private static Optional<BidType> bidTypeFromMtype(Integer mType) {
        return Optional.ofNullable(switch (mType) {
            case 1 -> BidType.banner;
            case 2 -> BidType.video;
            case null, default -> null;
        });
    }

    private static Optional<BidType> bidTypeFromExt(ObjectNode bidExt) {
        return Optional.ofNullable(bidExt)
                .map(ext -> ext.get("mediaType"))
                .filter(JsonNode::isTextual)
                .map(JsonNode::asText)
                .map(String::toLowerCase)
                .map(mediaType -> switch (mediaType) {
                    case "banner" -> BidType.banner;
                    case "video" -> BidType.video;
                    default -> null;
                });
    }

    private static Optional<BidType> bidTypeFromAdm(String adm) {
        if (StringUtils.isBlank(adm)) {
            return Optional.empty();
        }

        final String trimmed = adm.trim().toUpperCase();
        return trimmed.startsWith("<VAST") || trimmed.startsWith("<?XML")
                ? Optional.of(BidType.video)
                : Optional.empty();
    }

    private static Optional<BidType> bidTypeFromImp(String impId, List<Imp> imps) {
        for (Imp imp : imps) {
            if (!Objects.equals(imp.getId(), impId)) {
                continue;
            }

            final boolean hasBanner = imp.getBanner() != null;
            final boolean hasVideo = imp.getVideo() != null;
            if (hasVideo && !hasBanner) {
                return Optional.of(BidType.video);
            }
            if (hasBanner) {
                return Optional.of(BidType.banner);
            }

            break;
        }

        return Optional.empty();
    }
```

File: src/main/java/org/prebid/server/proto/openrtb/ext/request/revantage/ExtImpRevantage.java [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.Value;
+
+/**
+ * Bidder-specific portion of imp.ext.bidder for the Revantage adapter.
+ *
+ * <p>{@code feedId} is required. {@code placementId} and {@code publisherId}
+ * are optional pass-through identifiers.
+ */
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/proto/openrtb/ext/request/revantage/ExtImpRevantage.java:11]: no need for this comment

File: src/main/java/org/prebid/server/spring/config/bidder/RevantageConfiguration.java [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+
+@Configuration
+@PropertySource(value = "classpath:/bidder-config/revantage.yaml", factory = YamlPropertySourceFactory.class)
+public class RevantageConfiguration {
+
+    private static final String BIDDER_NAME = "revantage";
+
+    @Bean("revantageConfigurationProperties")
+    @ConfigurationProperties("adapters.revantage")
+    @Validated
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/spring/config/bidder/RevantageConfiguration.java]: No need for `@Validated`

**Review by CTMBNara — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+                final String feedId = StringUtils.trimToNull(ext.getFeedId());
+                if (feedId == null) {
+                    throw new PreBidException("imp %s: missing required param feedId".formatted(imp.getId()));
+                }
+                final Imp rewrittenImp = rewriteImpExt(imp, feedId, ext);
+                impsByFeed.computeIfAbsent(feedId, k -> new ArrayList<>()).add(rewrittenImp);
+            } catch (PreBidException e) {
+                errors.add(BidderError.badInput(e.getMessage()));
+            }
+        }
```
Comment (bukrain) [inline src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java:71]: nitpick 
I think we don't need to throw an exception here, we could just add to the errors list, when we don't have the feedId.
Something like

```suggestion
                final ExtImpRevantage ext = parseImpExt(imp);
                final String feedId = StringUtils.trimToNull(ext.getFeedId());
                if (feedId == null) {
                    errors.add(BidderError.badInput("imp %s: missing required param feedId".formatted(imp.getId())));
                    continue;
                }
                final Imp rewrittenImp = rewriteImpExt(imp, feedId, ext);
                impsByFeed.computeIfAbsent(feedId, k -> new ArrayList<>()).add(rewrittenImp);
```

File: src/test/java/org/prebid/server/bidder/revantage/RevantageBidderTest.java [hunk 14]
```diff
@@ -0,0 +1,333 @@
...
+        // given
+        final BidderCall<BidRequest> httpCall = givenHttpCall(204, null);
+
+        // when
+        final Result<List<BidderBid>> result = target.makeBids(httpCall, BidRequest.builder().build());
+
+        // then
+        assertThat(result.getErrors()).isEmpty();
+        assertThat(result.getValue()).isEmpty();
+    }
```
Comment (bukrain) [inline src/test/java/org/prebid/server/bidder/revantage/RevantageBidderTest.java:141]: As null is expected in the body, maybe it would make sense to first check if it's not null, since this test, doesn't seem to pass. 
Also can the body of the response be null when makeBids is called? Would expect for it to be an empty string ```""```.
@CTMBNara what do you think?

File: src/test/java/org/prebid/server/bidder/revantage/RevantageBidderTest.java [hunk 15]
```diff
@@ -0,0 +1,333 @@
...
+        final Result<List<BidderBid>> result = target.makeBids(httpCall, request);
+
+        // then
+        assertThat(result.getErrors()).isEmpty();
+        assertThat(result.getValue()).hasSize(1)
+                .extracting(BidderBid::getType).containsExactly(BidType.video);
+    }
+
+    @Test
+    public void makeBidsShouldDetectVideoFromVastWhenMtypeAbsent() throws Exception {
```
Comment (bukrain) [inline src/test/java/org/prebid/server/bidder/revantage/RevantageBidderTest.java:206]: nitpick
I would suggest to add a test cases for the adm that starts with ```<?XML``` and also for the case that adm is empty that it would be marked as an mtype of banner.

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java (hunk 4 above)
Comment (osulzhenko) [reply src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java]: @v0idxyz actual

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java (hunk 3 above)
Comment (osulzhenko) [reply src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java]: @v0idxyz actual

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java (hunk 5 above)
Comment (osulzhenko) [reply src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java]: @v0idxyz actual

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java (hunk 6 above)
Comment (osulzhenko) [reply src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java]: @v0idxyz actual

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java (hunk 7 above)
Comment (osulzhenko) [reply src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java]: @v0idxyz actual

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java (hunk 8 above)
Comment (osulzhenko) [reply src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java]: @v0idxyz actual

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java (hunk 9 above)
Comment (osulzhenko) [reply src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java]: @v0idxyz actual

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java (hunk 10 above)
Comment (osulzhenko) [reply src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java]: @v0idxyz actual

File: src/main/java/org/prebid/server/spring/config/bidder/RevantageConfiguration.java (hunk 12 above)
Comment (osulzhenko) [reply src/main/java/org/prebid/server/spring/config/bidder/RevantageConfiguration.java]: @v0idxyz actual

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java (hunk 3 above)
Comment (v0idxyz) [reply src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java]: @CTMBNara @osulzhenko Should be fixed now!

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java (hunk 4 above)
Comment (v0idxyz) [reply src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java]: @CTMBNara @osulzhenko Should be fixed now!

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java (hunk 5 above)
Comment (v0idxyz) [reply src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java]: @CTMBNara @osulzhenko Should be fixed now!

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java (hunk 6 above)
Comment (v0idxyz) [reply src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java]: @CTMBNara @osulzhenko Should be fixed now!

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java (hunk 7 above)
Comment (v0idxyz) [reply src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java]: @CTMBNara @osulzhenko Should be fixed now!

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java (hunk 8 above)
Comment (v0idxyz) [reply src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java]: @CTMBNara @osulzhenko Should be fixed now!

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java (hunk 10 above)
Comment (v0idxyz) [reply src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java]: @CTMBNara @osulzhenko Should be fixed now!

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java (hunk 9 above)
Comment (v0idxyz) [reply src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java]: @CTMBNara @osulzhenko Should be fixed now!

File: src/main/java/org/prebid/server/spring/config/bidder/RevantageConfiguration.java (hunk 12 above)
Comment (v0idxyz) [reply src/main/java/org/prebid/server/spring/config/bidder/RevantageConfiguration.java]: @CTMBNara @osulzhenko Removed

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java (hunk 13 above)
Comment (v0idxyz) [reply src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java:71]: @bukrain Should be fixed now!

**Comment by v0idxyz** [comment]
Please let me know if anything else has been missed or if we're good to go

**Review by bukrain — APPROVED** [review]
LGTM

**Comment by v0idxyz** [comment]
Any Info on a merge?

**Comment by osulzhenko** [comment]
@v0idxyz currently, PBS is on hold for adapter merges. The main focus is the next major update, 4.0, which should be finished by the end of Q2. Once that is done, we should be unblocked for merging other PRs. Sorry for the inconvenience

**Comment by v0idxyz** [comment]
> @v0idxyz currently, PBS is on hold for adapter merges. The main focus is the next major update, 4.0, which should be finished by the end of Q2. Once that is done, we should be unblocked for merging other PRs. Sorry for the inconvenience

Does this affect only PBS Java or PBS Go aswell?

File: src/main/java/org/prebid/server/proto/openrtb/ext/request/revantage/ExtImpRevantage.java (hunk 11 above)
Comment (CTMBNara) [reply src/main/java/org/prebid/server/proto/openrtb/ext/request/revantage/ExtImpRevantage.java:11]: ^^

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+    private HttpRequest<BidRequest> buildHttpRequest(BidRequest request, String feedId) {
+        final String uri = endpointUrl + "?feed=" + HttpUtil.encodeUrl(feedId);
+        return BidderUtil.defaultRequest(request, HttpUtil.headers(), uri, mapper);
+    }
+
+    @Override
+    public Result<List<BidderBid>> makeBids(BidderCall<BidRequest> httpCall, BidRequest bidRequest) {
+        if (httpCall.getResponse().getStatusCode() == HttpResponseStatus.NO_CONTENT.code()) {
+            return Result.empty();
+        }
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java:120]: No need for this check

File: src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+                continue;
+            }
+            final boolean hasBanner = imp.getBanner() != null;
+            final boolean hasVideo = imp.getVideo() != null;
+            if (hasVideo && !hasBanner) {
+                return Optional.of(BidType.video);
+            }
+            if (hasBanner) {
+                return Optional.of(BidType.banner);
+            }
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/revantage/RevantageBidder.java:200]: Can be simplified (missed last time):
```
            if (imp.getBanner() != null) {
                return Optional.of(BidType.banner);
            }
            if (imp.getVideo() != null) {
                return Optional.of(BidType.video);
            }
```

File: src/main/resources/bidder-config/revantage.yaml [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+        - video
+      site-media-types:
+        - banner
+        - video
+      supported-vendors:
+      vendor-id: 0
+    usersync:
+      redirect:
+        url: "https://sync.revantage.io/pbs/usersync?gdpr={{gdpr}}&gdpr_consent={{gdpr_consent}}&us_privacy={{us_privacy}}&gpp={{gpp}}&gpp_sid={{gpp_sid}}&r={{redirect_url}}"
+        uid-macro: "$UID"
```
Comment (CTMBNara) [inline src/main/resources/bidder-config/revantage.yaml:19]: Add empty line at the end of the file

File: src/test/java/org/prebid/server/bidder/revantage/RevantageBidderTest.java [hunk 19]
```diff
@@ -0,0 +1,333 @@
...
+    public void creationShouldFailOnInvalidEndpoint() {
+        assertThatIllegalArgumentException()
+                .isThrownBy(() -> new RevantageBidder("invalid_url", jacksonMapper));
+    }
+
+    @Test
+    public void makeHttpRequestsShouldReturnErrorWhenFeedIdMissing() {
+        // given
+        final BidRequest request = givenBidRequest(impBuilder ->
+                impBuilder.id("imp-1").ext(givenImpExt(null, null, null)));
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/revantage/RevantageBidderTest.java:55]: `impBuilder` - > `imp`

File: src/test/java/org/prebid/server/bidder/revantage/RevantageBidderTest.java [hunk 20]
```diff
@@ -0,0 +1,333 @@
...
+        assertThat(result.getErrors().get(0).getType()).isEqualTo(BidderError.Type.bad_input);
+        assertThat(result.getErrors().get(0).getMessage()).contains("feedId");
+    }
+
+    @Test
+    public void makeHttpRequestsShouldEmitOneRequestPerFeedId() {
+        // given
+        final Imp imp1 = givenImp(b -> b.id("imp-1").ext(givenImpExt("feed-one", null, null)));
+        final Imp imp2 = givenImp(b -> b.id("imp-2").ext(givenImpExt("feed-two", null, null)));
+        final Imp imp3 = givenImp(b -> b.id("imp-3").ext(givenImpExt("feed-one", null, null)));
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/revantage/RevantageBidderTest.java:72]: `b` -> `imp`

File: src/test/java/org/prebid/server/bidder/revantage/RevantageBidderTest.java [hunk 21]
```diff
@@ -0,0 +1,333 @@
...
+        assertThat(result.getValue().get(1).getUri()).isEqualTo(ENDPOINT_URL + "?feed=feed-two");
+        assertThat(result.getValue().get(1).getPayload().getImp())
+                .extracting(Imp::getId).containsExactly("imp-2");
+    }
+
+    @Test
+    public void makeHttpRequestsShouldRewriteImpExtToEndpointShape() {
+        // given
+        final BidRequest request = givenBidRequest(b ->
+                b.id("imp-1").ext(givenImpExt("feed-abc", "plc-1", "pub-1")));
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/revantage/RevantageBidderTest.java:96]: `b` -> `imp`

File: src/test/java/org/prebid/server/bidder/revantage/RevantageBidderTest.java [hunk 22]
```diff
@@ -0,0 +1,333 @@
...
+        assertThat(rewrittenExt.get("feedId").asText()).isEqualTo("feed-abc");
+        assertThat(rewrittenExt.get("bidder").get("placementId").asText()).isEqualTo("plc-1");
+        assertThat(rewrittenExt.get("bidder").get("publisherId").asText()).isEqualTo("pub-1");
+    }
+
+    @Test
+    public void makeHttpRequestsShouldUrlEncodeFeedIdInQueryParam() {
+        // given
+        final BidRequest request = givenBidRequest(b ->
+                b.id("imp-1").ext(givenImpExt("feed with spaces & symbols", null, null)));
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/revantage/RevantageBidderTest.java:119]: same

File: src/test/java/org/prebid/server/bidder/revantage/RevantageBidderTest.java [hunk 23]
```diff
@@ -0,0 +1,333 @@
...
+                                .id("b1").impid("imp-1").price(java.math.BigDecimal.valueOf(1.25))
+                                .adm("<div>Ad</div>").crid("c1").w(300).h(250).mtype(1)
+                                .build()))
+                        .build()))
+                .build();
+        final BidderCall<BidRequest> httpCall =
+                givenHttpCall(200, mapper.writeValueAsString(response));
+
+        final BidRequest request = givenBidRequest(b ->
+                b.id("imp-1").ext(givenImpExt("feed-abc", null, null)));
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/revantage/RevantageBidderTest.java:160]: `b` -> `imp`

File: src/test/java/org/prebid/server/bidder/revantage/RevantageBidderTest.java [hunk 24]
```diff
@@ -0,0 +1,333 @@
...
+        final BidderCall<BidRequest> httpCall =
+                givenHttpCall(200, mapper.writeValueAsString(response));
+
+        final BidRequest request = BidRequest.builder()
+                .imp(singletonList(Imp.builder()
+                        .id("imp-v")
+                        .video(Video.builder().w(640).h(480).build())
+                        .ext(givenImpExt("feed-v", null, null))
+                        .build()))
+                .build();
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/revantage/RevantageBidderTest.java:194]: use `givenBidRequest`

File: src/test/java/org/prebid/server/bidder/revantage/RevantageBidderTest.java (hunk 15 above)
Comment (CTMBNara) [reply src/test/java/org/prebid/server/bidder/revantage/RevantageBidderTest.java:206]: Add missed tests

File: src/test/java/org/prebid/server/bidder/revantage/RevantageBidderTest.java [hunk 25]
```diff
@@ -0,0 +1,333 @@
...
+
+        // when
+        final Result<List<BidderBid>> result = target.makeBids(httpCall, request);
+
+        // then
+        assertThat(result.getValue()).hasSize(1);
+        assertThat(result.getValue().get(0).getBidCurrency()).isEqualTo("USD");
+    }
+
+    // ------------ helpers ------------
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/revantage/RevantageBidderTest.java:308]: Remove comment

**Review by CTMBNara — CHANGES_REQUESTED** [review]

**Comment by osulzhenko** [comment]
@v0idxyz any update on this one?
